---
title: "INSITE README Guide"
author: "Nick Hoffman"
date: "September 2024"
output:
  html_document:
    df_print: paged
    highlight: pygment
    keep_md: no
    toc: true
    number_sections: true
    toc_depth: 1
    toc_float: true
    theme: journal
editor_options:
    chunk_output_type: inline
---

<style type="text/css">
h2, h3, h4, h5, h6 {
  counter-reset: section;
}
p {
  font-size:18px;
}

ul {
  font-size:18px;
}

li {
  font-size:18px;
}
table {
   padding: 0;border-collapse: collapse;
   layout: fixed;
   width: 90%; }
table tr {
   border-top: 1px solid #cccccc;
   background-color: white;
   margin: 0;
   padding: 0; }
table tr:nth-child(2n) {
   background-color: #f8f8f8; }
table tr th {
   font-weight: bold;
   border: 1px solid #cccccc;
   margin: 0;
   padding: 6px 13px; }
table tr td {
   border: 1px solid #cccccc;
   margin: 0;
   padding: 6px 13px; }
table tr th :first-child, table tr td :first-child {
   margin-top: 0; }
table tr th :last-child, table tr td :last-child {
   margin-bottom: 0; }
.html-widget {
    margin: auto;
}
</style>

---

```{r, results='asis', echo = F}
toc_depth <- rmarkdown::metadata$output$html_document$toc_depth
sel <- paste0("h",(toc_depth+1):10, collapse = " > span, ")
cat(paste0("<style>",
           sel, 
           " > .header-section-number { display: none; } </style>"))
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,comment=NA)
```

# Introduction

This guide will provide instructions on writing a README for data produced by INSITE-affiliated researchers.

# Why should I write a README?

INSITE is a collaborative project. In order for your collaborators to use your data easily and effectively, they will need to know the metadata which you provide in your README file. As well, writing a README is a requirement for upload to the INSITE server. Uploading to the server increases the security of your hard-won data. If something happens to your local copy of your data, you can retrieve it from the backup copy on the server. Lastly, journals often require that bioinformatic data be uploaded to specialist repositories like GenBank prior to publishing on those data. In turn, those specialist repositories require metadata which you provide in your README. Record the information as soon as you can, while it is still fresh in your memory.

# When do I write a README?

If you are producing any raw data with or without downstream products (final datasets, manuscripts, etc.), you should be uploading those data and any downstream products to the INSITE server. As well, you should be writing a README that provides a rich description of what those data are.

# What level do I write the README at?

The README should be written at the level of the experiment, not the individual file. So, for instance, if you've produced a transcriptome of an aster leafhopper population under heat stress, and it's the third version of this experiment that you've run, you might put your fastq files in the following subfolder:

```{r directory-structure,include=TRUE,echo=FALSE}
cat(paste("insiteData / data / leafhoppers / theme1 / transcriptome / exp3 "))

```

And you would write a README to be placed in the same folder with the following name:


```{r directory-structure5,include=TRUE,echo=FALSE}
cat(paste("[lastName]_[dateCreated]_[experiment#]_README.txt"))

```

Where experiment# would be replaced with a 3, since this is the README describing the files produced for your third experiment. This README will provide necessary metadata on your fastq files as well as any downstream data products (e.g., a .csv file that documents differential expression between your heat stressed leafhoppers and a control population) <b>Your README should explain what distinguishes this third version of the transcriptome experiment from others you've performed. </b>

# What information do I include in my README?

All READMEs <b>must</b> contain the following information (list adapted from <a href="https://data.research.cornell.edu/data-management/sharing/readme/" target="_blank">here</a>).

## General
* Dataset Title
* Name/institution/address/email information for person responsible for collecting the data
* List of files described by your README
* Date of data collection (single date or range)
* Information about geographic location of data collection

## Methodological information

* Description of methods for data collection/generation (include links or references to publications or other documentation containing experimental design or protocols used)
* Name of the species/strain being investigated (e.g., Aiptasia-CC7-Endo)
* Description of methods used for data processing (how were data generated from raw)
* Temperature at which experiment was performed if relevant
* Description of and links to computational scripts and software (r-code, python, command-line wrappers, etc.), as appropriate


## Access
* Any licenses or restrictions placed on data


## For each file:

* For each filename, short description of what data it contains
* Relationship among the files
* Date file created
* Meaning of the index number in file name - what distinguishes this file from a similarly named file?

### Data-specific information

* Variable list, including full names and definitions (spell out abbreviated words) of column headings for tabular (i.e., table) data
* Units of measurement for each variable
* Definitions of codes or symbols used to record missing or otherwise compromised data

# Extra requirements for sequencing data

Repositories of sequencing data like GenBank have specific metadata requirements for upload. If you are uploading sequencing data to the INSITE server, you must find the <a target="_blank" href="https://ncbi.nlm.nih.gov/biosample/docs/packages/">relevant metadata specification</a> from NCBI and upload a copy of the relevant spreadsheet (downloadable from NCBI) with all the mandatory fields filled in. This spreadsheet should be named as follows:


```{r directory-structure3,include=TRUE,echo=FALSE}
cat(paste("[lastName]_[dateCreated]_[experiment#]_[minInfoStandard].csv"))

```

For instance, if you're using the MIMARKS standard, you might name this file

```{r directory-structure4,include=TRUE,echo=FALSE}
cat(paste("Hoffman_09272024_3_MIMARKS.csv"))

```


<a href="https://www.ncbi.nlm.nih.gov/biosample/docs/packages/MIMARKS.specimen.host-associated.5.0/" target="_blank">MIMARKS</a> as specified by NCBI has the following requirements:

* sample name
* organism
* collection date
* broad scale environment
* local scale environment
* environmental medium
* geolocation name
* isolation and growth conditions
* latitude and longitude, 

as well as at least one of 

* strain
* isolate
* cultivar, or
* ecotype, 

and at least one of 

* host, or
* lab host.

These metadata values <b>must</b> be findable in a spreadsheet (.csv file) that you upload along with your README. The definitions of these metadata fields can be found on the relevant NCBI page. For MIMARKS, that would be <a href="https://ncbi.nlm.nih.gov/biosample/docs/packages/MIMARKS.specimen.host-associated.6.0/" target="_blank">here</a>.

# READMEs for collections of scripts

If you have a collection of coding scripts - in a GitHub repository, for instance - you may benefit from GitHub's guidance on writing READMEs, which you can find <a href="https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-readmes" target="_blank" > here</a>.
